### ISSUE TITLE
Darknet-19 YOLO

### ISSUE THREAD

**Issue by guantinglin** [eric612/MobileNet-YOLO#2]
Hi, 

Have you ever tried the darknet-19 version of yolov2? (original version in paper) 
When I replace the mobilenet into darknet-19, I can only get 55.5 mAP, which is even worse than yolov1...

**Comment by eric612** [comment]
Hi ,
Thank for your feedback , I'll try it next week , here is my guess  : 

1.  to get a good pre-train weights   
2.  solver type change to sgd

**Comment by eric612** [comment]
I first try to train original model and using pretrain coco weights (conver from original website , [tutorial](https://github.com/eric612/Caffe-YOLOv2-Windows) ), and mAP was 0.72 , deploy model just [uploaded](https://drive.google.com/file/d/1ZwANVrCp_gU5ow6Kx082pGWDLhNLVYC4/view?usp=sharing) , there are some reasons mAP be lower than original yolo : 

1. Original concat layer fixed shape when layer setup , so training can not use adaptive image size, like 320~608
2. I can't convert voc weights to caffemodel , I think there are some different between these frameworks
3. etc ..

And if have time and machine , I'll try to optimize it.

**Comment by guantinglin** [comment]
Hi,

Thanks for your reply. 
I have tried some modifications on the darknet19+yolov2 architecture.
Can you briefly describe how to test the mAP after training? 
I used the following command but only returned some weird information...
build/tools/caffe test --model yolo_darknet_test.prototxt --weights yolo_darknet_solver_iter_80000.caffemodel --iterations 4952 --gpu 0

**Comment by eric612** [comment]
I usually use to set solver max_iter  = 0 , and start training . It should start to test 
Note : do not modify batch size of test prototxt , currently , the yolo eval function do not support batch size > 1

**Comment by guantinglin** [comment]
Actually, I have tried so...
But got no response...
![image](https://user-images.githubusercontent.com/25660585/42485816-d8c561de-842b-11e8-99ab-b786b239e150.png)

Here is my solver setting
![image](https://user-images.githubusercontent.com/25660585/42485853-0953c232-842c-11e8-99e3-d60cf02abd47.png)

**Comment by eric612** [comment]
Below is my solver setting , and it works  , 
I think if it do not show mAP , Maybe eval layer didn't receive any detection results . It happened when used mismatch prototxt and model 

train_net: "models/darknet/darknet19_train.prototxt"
test_net: "models/darknet/darknet19_test.prototxt"
test_iter: 4952
test_interval: 1000
base_lr: 0.001
display: 10
max_iter: 0
lr_policy: "multistep"
gamma: 0.1
weight_decay: 0.00005
snapshot: 1000
snapshot_prefix: "models/darknet/yolov2_deploy"
solver_mode: GPU
debug_info: false
snapshot_after_train: true
test_initialization: false
average_loss: 10
stepvalue: 60000
stepvalue: 80000
iter_size: 2
type: "SGD"
eval_type: "detection"
ap_version: "11point"
show_per_class_result: true

script : 
./tools/caffe train --solver models/darknet/darknet19_solver.prototxt --weights models/darknet/yolov2_deploy_iter_50000.caffemodel --gpu=0 2>&1 | tee $LOG

![image](https://user-images.githubusercontent.com/15793472/42486847-ab56bc7a-8430-11e8-88a2-31abbc78b8f7.png)

**Comment by eric612** [comment]
Sorry , I forgot to update test.prototxt

**Comment by guantinglin** [comment]
Hi, 

I trained the darknet-yolov2 with the pre-trained weights(darknet-19, 448 in imagenet, not coco) released [here](https://pjreddie.com/darknet/imagenet/#darknet53), and convert to caffe format with this [repo](https://github.com/marvis/pytorch-caffe-darknet-convert).
I have tried many kinds of parameter setting but the best mAP that I can reach is less than 60%, which is much lower than 76%. I have sent you the model to you via your gmail, can you help inspect what is going wrong in my training procedure?

**Comment by eric612** [comment]
I'm not sure , but I don't have machine can try to training now , and below is my guess : 

1. Transform param in test.prototxt was
    scale: 0.007843
    mean_value: 127.5
    mean_value: 127.5
    mean_value: 127.5

   But in the train.prototxt was gone in your prototxt , and I suggest scale set to 1/255.5 and mean set to 
   zero

2. I try to convert darknet.cov23 to caffemodel yet , but count was dismatch, I didn't solve this problem , 
    and I can not ensure the caffemodel was correct

**Comment by guantinglin** [comment]
Hi,

Sorry to bother you again...

1. I found that I sent you the wrong train.prototxt. It does have transform param settings in my experiments, actually. The darknet pre-processing scales the pixel value into [-1,+1], so I think it is right to set the transform param as:
    scale: 0.007843 
    mean_value: 127.5
    mean_value: 127.5
    mean_value: 127.5

2. I think the converted weights with this [repo](https://github.com/marvis/pytorch-caffe-darknet-convert) is correct for I have tried the converted weights for other architectures and it works.

Thank you
